### PR TITLE
Validate contains change

### PR DESCRIPTION
**Pull request by markiantorno**
Give me a moment to review this myself first, then  I will provide details and tag people.

### CODE REVIEW

**Comment by markiantorno** [comment]
Alright so, these changes are related to bumping the dependency on [org.hl7.fhir.core](https://github.com/hapifhir/org.hl7.fhir.core) to version 5.3.0. 

The main updates that resulted in the majority of the changes in this PR were the addition of the interface [IValidationPolicyAdvisor](https://github.com/hapifhir/org.hl7.fhir.core/blob/4d0692ff61880f37083ad0738fcf25571dcc9cc5/org.hl7.fhir.r5/src/main/java/org/hl7/fhir/r5/utils/validation/IValidationPolicyAdvisor.java), and the refactoring of some of the main validation classes (for cleaning purposes). Also, many of the error messages were updated, so the resulting tests in HAPI had to be modified to reflect these message changes.

In regard to the new interface IValidationPolicyAdvisor, by implementing this interface and passing it through the validation context in HAPI to the [InstanceValidator] (https://github.com/hapifhir/org.hl7.fhir.core/blob/master/org.hl7.fhir.validation/src/main/java/org/hl7/fhir/validation/instance/InstanceValidator.java) in the core libraries, users can now control the behavior of the validator when validating references, contained resources, and coded content. Previously, only references were controlled in this way, and users controlled this by overriding the validation fetcher.

Test cases were added in the [validation test cases repository](https://github.com/FHIR/fhir-test-cases/) to demonstrate it's functionality. In particular, the two test cases `contained-resource-bad-id`, and `contained-resource-bad-id-ignore`. The definitions and expected outcomes of these test cases are in the [manifest.json](https://github.com/FHIR/fhir-test-cases/blob/master/validator/manifest.json) in that repository.

Please let me know if you have any questions.

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/validation/FhirValidator.java [hunk 1]
```diff
@@ -52,7 +52,7 @@
 	private static volatile Boolean ourPhPresentOnClasspath;
 	private final FhirContext myContext;
 	private List<IValidatorModule> myValidators = new ArrayList<>();
-	private IInterceptorBroadcaster myInterceptorBraodcaster;
+	private IInterceptorBroadcaster iInterceptorBroadcaster;
```
Comment (tadgh) [inline hapi-fhir-base/src/main/java/ca/uhn/fhir/validation/FhirValidator.java:55]: HAPI convention is `my` as a prefix for member variables.

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/dstu3/FhirResourceDaoDstu3ValidateTest.java [hunk 2]
```diff
@@ -121,14 +121,14 @@ public void testExpandLocalCodeSystemWithExplicitCodes() throws IOException {
...
-		val.setBestPracticeWarningLevel(org.hl7.fhir.r5.utils.IResourceValidator.BestPracticeWarningLevel.Warning);
+		val.setBestPracticeWarningLevel(BestPracticeWarningLevel.Warning);
 	}
 
 
 	@Test
 	public void testValidateWithCanonicalReference() {
 		FhirInstanceValidator val = AopTestUtils.getTargetObject(myValidatorModule);
-		org.hl7.fhir.r5.utils.IResourceValidator.BestPracticeWarningLevel a = IResourceValidator.BestPracticeWarningLevel.Ignore;
+		BestPracticeWarningLevel a = BestPracticeWarningLevel.Ignore;
```
Comment (tadgh) [inline hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/dstu3/FhirResourceDaoDstu3ValidateTest.java:131]: may as well fix the variable name while you're in here. `a` is not great.

File: hapi-fhir-validation/src/main/java/org/hl7/fhir/common/hapi/validation/validator/FhirDefaultPolicyAdvisor.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+	@Override
+	public ContainedReferenceValidationPolicy policyForContained(IResourceValidator validator,
+																					 Object appContext,
+																					 String containerType,
+																					 String containerId,
+																					 Element.SpecialElement containingResourceType,
+																					 String path,
+																					 String url) {
+		return ContainedReferenceValidationPolicy.CHECK_VALID;
```
Comment (tadgh) [inline hapi-fhir-validation/src/main/java/org/hl7/fhir/common/hapi/validation/validator/FhirDefaultPolicyAdvisor.java:38]: this spacing kinda wild. might just be GH formatting, but try to de-dent the parameters a few levels.

File: hapi-fhir-validation/src/main/java/org/hl7/fhir/common/hapi/validation/validator/VersionTypeAdvisorDstu21.java [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package org.hl7.fhir.common.hapi.validation.validator;
+
+import org.hl7.fhir.convertors.advisors.impl.BaseAdvisor_14_50;
+
+public class VersionTypeAdvisorDstu21 extends BaseAdvisor_14_50 {
```
Comment (tadgh) [inline hapi-fhir-validation/src/main/java/org/hl7/fhir/common/hapi/validation/validator/VersionTypeAdvisorDstu21.java:5]: 14_50 ? what is 14? and what does 21 mean here?

File: hapi-fhir-validation/src/test/java/org/hl7/fhir/r4/validation/FhirInstanceValidatorR4Test.java [hunk 5]
```diff
@@ -1396,14 +1402,17 @@ public void testInvocationOfValidatorFetcher() throws IOException {
...
+		when(policyAdvisor.policyForReference(any(), any(), any(), any())).thenReturn(ReferenceValidationPolicy.CHECK_TYPE_IF_EXISTS);
+		when(policyAdvisor.policyForContained(any(), any(), any(), any(), any(), any(), any())).thenReturn(ContainedReferenceValidationPolicy.CHECK_TYPE);
 		myInstanceVal.setValidatorResourceFetcher(resourceFetcher);
+		myInstanceVal.setValidatorPolicyAdvisor(policyAdvisor);
 		myVal.validateWithResult(encoded);
 
-		verify(resourceFetcher, times(15)).resolveURL(any(), any(), anyString(), anyString(), anyString());
-		verify(resourceFetcher, times(12)).validationPolicy(any(), any(), anyString(), anyString());
-		verify(resourceFetcher, times(12)).fetch(any(), any(), anyString());
+		verify(resourceFetcher, times(12)).resolveURL(any(), any(), anyString(), anyString(), anyString());
```
Comment (tadgh) [inline hapi-fhir-validation/src/test/java/org/hl7/fhir/r4/validation/FhirInstanceValidatorR4Test.java:1414]: any clue why this number had to change? Just internal refactoring requiring less calls?

**Review by tadgh — CHANGES_REQUESTED** [review]
Overall looks good, just a couple minor questions for my own clarity, and some minor style stuff. 

1. This requires a changelog
2. Does this change _break_ Smile CDR? If so, bumping the HAPI version makes sense here, otherwise you can leave it at PRE3. 

Otherwise, lgtm.

File: hapi-fhir-validation/src/main/java/org/hl7/fhir/common/hapi/validation/validator/VersionTypeAdvisorDstu21.java (hunk 4 above)
Comment (markiantorno) [reply hapi-fhir-validation/src/main/java/org/hl7/fhir/common/hapi/validation/validator/VersionTypeAdvisorDstu21.java:5]: I didn't wite it,but I'm pretty sure DSTU21 is MAY2014 release that was a special release done for Telus.

It is referred to in the conversion libraries by the shortform '14'.
```
10 = DSTU2
14 = DSTU2014MAY
30 = DSTU3
40 = R4
50 = R5
```

So, if you can tell the purpose of a conversion class/util through it's name
***_10_50 -> Converts from DSTU2 to R5 and back
***_14_50 -> Converts from DSTU2014 to R5 and back
***_10_30 -> DSTU2 to DSTU3 and back
etc

File: hapi-fhir-validation/src/test/java/org/hl7/fhir/r4/validation/FhirInstanceValidatorR4Test.java (hunk 5 above)
Comment (markiantorno) [reply hapi-fhir-validation/src/test/java/org/hl7/fhir/r4/validation/FhirInstanceValidatorR4Test.java:1414]: The internal refactoring in the validator resulted in less calls. Same result as before though.

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/dstu3/FhirResourceDaoDstu3ValidateTest.java (hunk 2 above)
Comment (markiantorno) [reply hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/dstu3/FhirResourceDaoDstu3ValidateTest.java]: Fixed

File: hapi-fhir-validation/src/main/java/org/hl7/fhir/common/hapi/validation/validator/FhirDefaultPolicyAdvisor.java (hunk 3 above)
Comment (markiantorno) [reply hapi-fhir-validation/src/main/java/org/hl7/fhir/common/hapi/validation/validator/FhirDefaultPolicyAdvisor.java:31]: Fixed

**Comment by markiantorno** [comment]
> Overall looks good, just a couple minor questions for my own clarity, and some minor style stuff.
> 
>     1. This requires a changelog
> 
>     2. Does this change _break_ Smile CDR? If so, bumping the HAPI version makes sense here, otherwise you can leave it at PRE3.
> 


Chagelog added, and this will break things in CDR, so I bumped to PRE-4

**Review by tadgh — APPROVED** [review]

**Pull request merged**